### PR TITLE
refactor the code to allow overwriting the function to get config

### PR DESCRIPTION
**Pull request by iclementine**
1. Refactor the code the allow overwriting the function to get config;
2. Add a fall-back config（`BLOCK_M, BLOCK_N, num_stages, num_warps = 32, 32, 1, 4`） when no rules match the device and the input;
3. Add an example for using custom config(examples/use_cutom_config_func.py);
4. rename test cases 'test_attention_fwd_bwd' to 'test_attention_bwd' to allow better filtering with pytest.

### CODE REVIEW

File: src/flag_attn/piecewise.py [hunk 1]
```diff
@@ -21,31 +21,49 @@
...
+        Dq1, Dk1, Dq2, Dk2, Dv = q1.shape[-1], k1.shape[-1], q2.shape[-1], k2.shape[-1], v.shape[-1]
+        assert Dq1 == Dk1 == Dq2 == Dk2 == Dv
+        assert Dk1 in {16, 32, 64, 128}
+
+        B, H, M, D = q1.shape
+        N = k1.shape[2]
+        if sm_scale is None:
+            sm_scale = 1. / math.sqrt(D)
+        
+        # to work around https://github.com/openai/triton/issues/2441
```
Comment (StrongSpoon) [inline src/flag_attn/piecewise.py:38]: ditto

File: src/flag_attn/piecewise.py [hunk 2]
```diff
@@ -55,193 +73,94 @@ def forward(ctx, q1, k1, q2, k2, v, w, causal, sm_scale):
...
 
-        dq1, dk1, dq2, dk2, dv = standalone_backward(q1, k1, q2, k2, v, w, causal, sm_scale, o, L, do)
+        B, H, M, D = q1.shape
+        N = k1.shape[2]
+        P_SEQ = N - M
+
+        if sm_scale is None:
+            sm_scale = 1. / math.sqrt(D)
+
+        # to work around https://github.com/openai/triton/issues/2441
```
Comment (StrongSpoon) [inline src/flag_attn/piecewise.py:88]: ditto

File: src/flag_attn/piecewise.py [hunk 3]
```diff
@@ -21,31 +21,49 @@
...
-            w(int): The threshold of distance between q and k. When the distance is not greater than w, the attention score is dot(q1, k1); otherwise dot(q2, k2).
-            causal(bool): Whether causal masking is applied to attention scores before applying softmax.
-            sm_scale(float): The scaling of attention scores before applying softmax.
-
-        Returns:
-            out: (torch.Tensor): The output. The shape is (batch_size, nheads, seqlen_q, headdim).
-        """
-        o, L = standalone_forward(q1, k1, q2, k2, v, w, causal, sm_scale)
+        # shape constraints
+        Dq1, Dk1, Dq2, Dk2, Dv = q1.shape[-1], k1.shape[-1], q2.shape[-1], k2.shape[-1], v.shape[-1]
```
Comment (StrongSpoon) [inline src/flag_attn/piecewise.py:29]: 这里好像不太优雅

File: src/flag_attn/flash.py [hunk 4]
```diff
@@ -3,226 +3,181 @@
...
-            v.stride(0), v.stride(1), v.stride(2), v.stride(3),
-            o.stride(0), o.stride(1), o.stride(2), o.stride(3),
-            B, H, M, P_SEQ,
-            BLOCK_M=BLOCK_M, BLOCK_N=BLOCK_N, BLOCK_DMODEL=D, IS_CAUSAL=causal,
-            DIVISIBLE_M=divisible_m, DIVISIBLE_N=divisible_n, 
-            num_warps=num_warps, num_stages=num_stages,
-        )
-
+        
+        # to work around https://github.com/openai/triton/issues/2441
```
[the remark this answers is not in the document]
Comment (iclementine) [reply src/flag_attn/flash.py:23]: around?

File: src/flag_attn/piecewise.py (hunk 3 above)
Comment (iclementine) [reply src/flag_attn/piecewise.py:29]: 是变量名的问题还是 tuple 的赋值方式呢？

Comment (iclementine) [reply src/flag_attn/piecewise.py:29]: 是变量名的问题还是元组太长？

File: src/flag_attn/flash.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (iclementine) [reply src/flag_attn/flash.py:23]: That's what I mean. To avoid the problem without solving it.

File: src/flag_attn/piecewise.py (hunk 3 above)
Comment (iclementine) [reply src/flag_attn/piecewise.py:29]: 等之后我们引入 formatter 吧。

**Review by StrongSpoon — APPROVED** [review]

**Pull request merged**